diffusion_with_autoscaler/strategies.py

```diff
@@ -0,0 +1,151 @@
+        if len(keys) > 1:
+            selected_url = np.random.choice(keys, p=list(local_router_metadata.values()))
+        else:
+            selected_url = keys[0]
+        return selected_url, method
+
+    def make_request(self, request: Request, full_path: str, local_router_metadata: Any, payload) -> Response:
+
+        if self._session is None:
+            self._session = _configure_session()
+
+        selected_url, method = self.select_url(request, local_router_metadata)
+        if method == "post":
+            return getattr(self._session, method)(selected_url + "/" + full_path, json=payload)
+        else:
+            return getattr(self._session, method)(selected_url + "/" + full_path)
+
+    @abc.abstractmethod
+    def run(
+        self,
+        serve_works: List[LightningWork],
+        create_work: Callable,
+        replace_work: Callable,
+    ) -> Any:
+        """Controls the following:
+        - when to create works
+        - how to create works, e.g. create_work(weight_path="path/to/new_weights")
+        - when to replace old works with new works
+        """
+
+        pass
+
+    def on_after_run(self, serve_works: List[LightningWork], res):
+        pass
+
+
+class PreemptibleRollout(Strategy):
+    def __init__(self, interval: int = 30 * 60) -> None:
+        """
+        This strategy implements a mechanism to automatically replace servers on a scheduled internal
+        to continously run on spot instances.
+
+        Arguments:
+            interval: Time in seconds before creating a replacement server.
+
+        """
+        super().__init__()
+        self.interval = interval
+        self._work_start_tracker = {}
+        self._old_works = []
+        self.new_works = List()
+        self.fake_trigger = 0
+
+    def run(
+        self,
+        serve_works: List[LightningWork],
+        create_work: Callable,
+        replace_work: Callable,
+    ) -> None:
+        self.fake_trigger += 1  # Note: change state to keep calling `run`.
+        # print(self.fake_trigger)
+        # serve_works = [w for w in serve_works if w.cloud_compute.preemptible]
+
+        for old_work in serve_works:
+            if old_work.url and old_work not in self._work_start_tracker:
+                print(f"Tracking preemptive {old_work.name}.")
+                self._work_start_tracker[old_work] = time.time()
+
+        for old_work, start_time in self._work_start_tracker.items():
+            if self.interval < (time.time() - start_time) and old_work not in self._old_works:
+                new_work = create_work()
+                print(f"Created a new work {new_work}")
+                self.new_works.append(new_work)
+                print(f"Appended the new work {new_work} to new_works")
+                self._old_works.append(old_work)
+                print(f"Appended the old work {old_work} to old_works")
+
+        # spin up machines
+        for new_work in self.new_works:
```

Comment:
Yes, we would need to handle the logic when trying to replace. If the old_work was removed before new_work is ready, new_work should be deleted.